optimize/cmaes.go

```diff
@@ -0,0 +1,381 @@
+			panic("cma-es-chol: bad cholesky. shouldn't happen")
+		}
+		cma.chol = chol
+	}
+
+	cma.evals = make([]int, cma.pop)
+	for i := 0; i < cma.pop; i++ {
+		cma.evals[i] = i
+	}
+
+	cma.bestX = resize(cma.bestX, dim)
+	cma.bestF = math.Inf(1)
+
+	t := min(tasks, cma.pop)
+	cma.taskIdxs = make([]int, cma.pop)
+	for i := 0; i < t; i++ {
+		cma.taskIdxs[i] = -1
+	}
+	// Get a new waitgroup so that if the struct is re-used the.
+	cma.mux = sync.Mutex{}
+	cma.wg = sync.WaitGroup{}
+	return t
+}
+
+func (cma *CmaEsChol) IterateGlobal(task int, loc *Location) (Operation, error) {
+	// Check the status of the incoming task. If it is a number, it means
+	// that taks contains a valid location.
+	idx := cma.taskIdxs[task]
+	if idx != -1 {
+		cma.fs[idx] = loc.F
+		cma.wg.Done()
+	}
+
+	// Get the next task and send it to be run if there is a next task to be run.
+	// If all of the tasks have been run, perform an update step. Note that the
+	// use of this mutex means that only one task can proceed, all of the
+	// other tasks should get stuck and then get a new location.
+	cma.mux.Lock()
+	if len(cma.evals) != 0 {
+		// There are still tasks to evaluate. Grab one and remove it from the list.
+		newIdx := cma.evals[len(cma.evals)-1]
+		cma.evals = cma.evals[:len(cma.evals)-1]
+		cma.wg.Add(1)
+		cma.mux.Unlock()
+
+		// Sample x and send it to be evaluated.
+		distmv.NormalRand(cma.xs.RawRowView(newIdx), cma.mean, &cma.chol, cma.Src)
+		copy(loc.X, cma.xs.RawRowView(newIdx))
+		cma.taskIdxs[task] = newIdx
+		return FuncEvaluation, nil
+	}
+	// There are no more tasks to evaluate. This means the iteration is over.
+	// Find the best current f, update the parameters, and re-establish
+	// the evaluations to run.
+
+	// Wait for all of the outstanding tasks to finish, so the full set of functions
+	// has been evaluated.
+	cma.wg.Wait()
+
+	// Find the best f out of all the tasks.
+	best := floats.MinIdx(cma.fs)
+	bestF := cma.fs[best]
+	bestX := cma.xs.RawRowView(best)
+	if cma.ForgetBest {
+		loc.F = bestF
+		copy(loc.X, bestX)
+	} else {
+		if bestF < cma.bestF {
+			cma.bestF = bestF
+			copy(cma.bestX, bestX)
+		}
+		loc.F = cma.bestF
+		copy(loc.X, cma.bestX)
+	}
+
+	cma.taskIdxs[task] = -1
+
+	// Update the parameters of the distribution
+	err := cma.update()
+
+	// Reset the tasks
+	cma.evals = cma.evals[:cma.pop]
+
+	cma.mux.Unlock()
+	return MajorIteration, err
+}
+
+// update computes the new parameters (mean, cholesky, etc.)
+func (cma *CmaEsChol) update() error {
+	// Sort the function values to find the elite samples.
+	ftmp := make([]float64, cma.pop)
+	copy(ftmp, cma.fs)
+	indexes := make([]int, cma.pop)
+	for i := range indexes {
+		indexes[i] = i
+	}
+	sort.Sort(bestSorter{F: ftmp, Idx: indexes})
+
+	meanOld := make([]float64, len(cma.mean))
+	copy(meanOld, cma.mean)
+
+	// m_{t+1} = \sum_{i=1}^mu w_i x_i
+	for i := range cma.mean {
+		cma.mean[i] = 0
+	}
+	for i, w := range cma.weights {
+		idx := indexes[i] // index of teh 1337 sample.
```

Comment:
`s/teh/the/`

Reply:
I can change it if you'd like, but that one was part of the joke  (https://en.wikipedia.org/wiki/Leet#Vocabulary)

Reply:
h4! you c4n 134v3 as is.